magicblock-committor-service/src/intent_executor/two_stage_executor.rs

```diff
@@ -0,0 +1,221 @@
+use std::ops::{ControlFlow, Deref};
+
+use log::{error, info, warn};
+use solana_pubkey::Pubkey;
+
+use crate::{
+    intent_executor::{
+        error::{
+            IntentExecutorError, IntentExecutorResult,
+            TransactionStrategyExecutionError,
+        },
+        task_info_fetcher::TaskInfoFetcher,
+        ExecutionOutput, IntentExecutorImpl,
+    },
+    persist::IntentPersister,
+    tasks::task_strategist::TransactionStrategy,
+    transaction_preparator::TransactionPreparator,
+};
+
+pub struct TwoStageExecutor<'a, T, F> {
+    inner: &'a IntentExecutorImpl<T, F>,
+}
+
+impl<'a, T, F> TwoStageExecutor<'a, T, F>
+where
+    T: TransactionPreparator,
+    F: TaskInfoFetcher,
+{
+    pub fn new(executor: &'a IntentExecutorImpl<T, F>) -> Self {
+        Self { inner: executor }
+    }
+
+    pub async fn execute<P: IntentPersister>(
+        &self,
+        committed_pubkeys: &[Pubkey],
+        mut commit_strategy: TransactionStrategy,
+        mut finalize_srategy: TransactionStrategy,
+        junk: &mut Vec<TransactionStrategy>,
+        persister: &Option<P>,
+    ) -> IntentExecutorResult<ExecutionOutput> {
+        const RECURSION_CEILING: u8 = 10;
+
+        let mut i = 0;
+        let (commit_result, last_commit_strategy) = loop {
+            i += 1;
+
+            // Prepare & execute message
+            let execution_result = self
+                .prepare_and_execute_strategy(&mut commit_strategy, persister)
+                .await
+                .map_err(IntentExecutorError::FailedCommitPreparationError)?;
+            let execution_err = match execution_result {
+                Ok(value) => break (Ok(value), commit_strategy),
+                Err(err) => err,
+            };
+
+            let flow = self
+                .patch_commit_strategy(
+                    &execution_err,
+                    &mut commit_strategy,
+                    committed_pubkeys,
+                )
+                .await?;
+            let cleanup = match flow {
+                ControlFlow::Continue(value) => {
+                    info!("Patched intent, error was: {:?}", execution_err);
+                    value
+                }
+                ControlFlow::Break(()) => {
+                    break (Err(execution_err), commit_strategy)
+                }
+            };
+
+            if i >= RECURSION_CEILING {
+                error!(
```

Comment:
What happens then?
Does it ever get retried? It's not clear.
Ideally the intent should already be persisted to the DB and retried later?

I understand though that it might be blocking intents that follow and are dependent on this one.
How is this resolved?

The PR description doesn't go into detail either.